Tests/Marketplace/search_and_install_packs.py

```diff
@@ -168,8 +168,79 @@ def search_pack(client: demisto_client, prints_manager: ParallelPrintsManager, p
+        request_timeout (int): Timeout settings for the installation request.
+
+    Returns:
+
+    """
+    packs_to_install_str = ', '.join([pack['id'] for pack in packs_to_install])
+    message = 'Installing the following packs in server {}:\n{}'.format(host, packs_to_install_str)
+    prints_manager.add_print_job(message, print_color, thread_index, LOG_COLORS.GREEN, include_timestamp=True)
+    prints_manager.execute_thread_prints(thread_index)
+    # make the pack installation request
+    global PACK_INSTALL
+    PACK_INSTALL = False
+    request_data = {
+        'packs': packs_to_install,
+        'ignoreWarnings': True
+    }
+    while PACK_INSTALL is not True:
+        try:
+            response_data, status_code, _ = demisto_client.generic_request_func(client,
+                                                                                path='/contentpacks/marketplace/install',
+                                                                                method='POST',
+                                                                                body=request_data,
+                                                                                accept='application/json',
+                                                                                _request_timeout=request_timeout)
+
+            if 200 <= status_code < 300:
+                message = 'Packs were successfully installed!\n'
+                prints_manager.add_print_job(message, print_color, thread_index, LOG_COLORS.GREEN,
+                                             include_timestamp=True)
+            else:
+                result_object = ast.literal_eval(response_data)
+                message = result_object.get('message', '')
+                err_msg = f'Failed to install packs - with status code {status_code}\n{message}\n'
+                prints_manager.add_print_job(err_msg, print_error, thread_index, include_timestamp=True)
+                raise Exception(err_msg)
+
+            PACK_INSTALL = True
+            break
+
+        except Exception as e:
+            err_msg = f'The request to install packs has failed. Reason:\n{str(e)}\n'
+            prints_manager.add_print_job(err_msg, print_error, thread_index, include_timestamp=True)
+            PACK_INSTALL = False
+            pack_id = ''
+            message = str(e).split('\n')
+            # Get the pack ID of the failed pack.
+            for line in message:
+                if line.startswith('HTTP response body: '):
+                    error_message = json.loads(line.split(': ', 1)[1])
+                    error = error_message.get('error')
+                    pack_id = error.split()[-2]
+            # Removed the bad pack from the list
+            packs = [pack for pack in packs_to_install if not (pack['id'] == pack_id)]
+            request_data = {
+                'packs': packs,
+                'ignoreWarnings': True
+            }
```

Comment:
I think this logic should be in it's own method where you can document the voodoo that's being done here and unittest it.